### PR TITLE
New Feature : conflicting claims

### LINKED ISSUE #260: Add conflicting claims section on accept claims page
Based on the issue and pull request url, find the list of conflicting claims and display them on the accept claims page.

Solutions with better layout, and covering all the edge cases will be preferred.

Things to take care of:
1. If the target claim is for a pull request (i.e the pull request url point to a pr on github and not an issue), then only find conflicting claims which are also for a PR only. Some claims are added for suggesting features/issues, in that case the the pull request url point to an issue on github.
2. There can be terminating '/' in some urls and not others. Can be similar cases for 'http' and 'https'. Use the `like` sql operator to make db queries where ever required to cover such cases.

This issue has 700 bounty points.

Screenshot 2020-05-17 at 8 36 55 PM

### PR DESCRIPTION
**Pull request by YashKumarVerma**
Fixes #260 
Ref #258

- [x] Shows a list of conflicting claims made by various users in the sidebar
- [x] Provision to match URLs who differ only in the protocol, trailing slashes, ports, hashes, query strings etc 
- [x] Inline documentation added for codebase edits
- [x] All errors inside route handled

### Preview
- Raw data used to build page
screencapture-localhost-3232-claims-2-2020-05-18-12_40_48

- When there are no conflicts
Screenshot from 2020-05-18 13-38-02

- When conflicts arise
Screenshot from 2020-05-18 13-38-14

### CODE REVIEW

File: utils/datautils.js [hunk 1]
```diff
@@ -48,6 +48,18 @@ function getClaimById(claimId) {
   return db.Claim.findById(claimId)
 }
 
+function getConflictingClaimByPr(pullRequestId) {
```
Comment (hereisnaman) [inline utils/datautils.js:51]: Not the correct logic. just checking the presence of id does not decide the id is for pr or issue.

File: routes/root.js [hunk 2]
```diff
@@ -193,15 +193,52 @@ route.get('/claims/add', auth.ensureLoggedInGithub, (req, res) => {
...
-    .catch(err => {
-      res.send('Error fetching claim id = ' + escapeHtml(req.params.id))
+route.get('/claims/:id', auth.adminOnly, async (req, res) => {
+  try {
+    const claimInfo = await du.getClaimById(req.params.id)
+    if (!claimInfo) {
+      throw new Error('Error finding claim')
+    }
+
+    const prSlug = new URL(claimInfo.pullUrl).pathname.replace(/\/+$/, '')
```
Comment (hereisnaman) [inline routes/root.js:203]: Write a deserializer in utils to convert a url to:

{
project: "",
type: "pr" | "issue",
id: "",
}

File: routes/root.js [hunk 3]
```diff
@@ -193,15 +193,52 @@ route.get('/claims/add', auth.ensureLoggedInGithub, (req, res) => {
...
+    let response, conflicts
+
+    /**
+     * the database has internal checks for duplicate items in the pull request column
```
Comment (hereisnaman) [inline routes/root.js:207]: does not cover all the cases.

File: utils/datautils.js (hunk 1 above)
Comment (YashKumarVerma) [reply utils/datautils.js:51]: I think i named the parameter wrongly in this one.
for a the url
`https://github.com/coding-blocks/boss/pull/279`
this is the value of the parameter
`coding-blocks/boss/pull/279`
and not only `279`

File: routes/root.js (hunk 2 above)
Comment (YashKumarVerma) [reply routes/root.js:203]: working on it

File: routes/root.js (hunk 3 above)
Comment (YashKumarVerma) [reply routes/root.js:207]: could you please mention which type of case is not handled

File: utils/datautils.js (hunk 1 above)
Comment (YashKumarVerma) [reply utils/datautils.js:51]: image

**Comment by YashKumarVerma** [comment]
@hereisnaman waiting for review

File: utils/datautils.js (hunk 1 above)
Comment (YashKumarVerma) [reply utils/datautils.js]: This checks the unique part of the URL, which is of the type `user/repository/pull/id`. The above confusion arose due to improper variable naming, which I've fixed now.

**Comment by hereisnaman** [comment]
> @hereisnaman, @championswimmer waiting for review

Reviews usually take from 1-3 days. You can start working on other issues.

**Comment by YashKumarVerma** [comment]
Awaiting review.

**Comment by YashKumarVerma** [comment]
Requesting someone to carry it forward.

**Comment by YashKumarVerma** [comment]
A few thigns to mention:
- if only the number is matched using a wildcard, then issue 10 will show conflicts with 100, 101, 102...
- if it will be assumed that the issue / pr id will be followed by a slash or terminate immediately, it's also a wrong lead, as URLs ending with hashes are also submitted

Working on a solution of the above problems.

**Comment by hereisnaman** [comment]
> A few thigns to mention:
> 
>     * if only the number is matched using a wildcard, then issue 10 will show conflicts with 100, 101, 102...
> 
>     * if it will be assumed that the issue / pr id will be followed by a slash or terminate immediately, it's also a wrong lead, as URLs ending with hashes are also submitted
> 
> 
> Working on a solution of the above problems.

What is the use case behind having urls with hashes? Can't we clean them in the add claim form?

**Comment by YashKumarVerma** [comment]
> > A few thigns to mention:
> > ```
> > * if only the number is matched using a wildcard, then issue 10 will show conflicts with 100, 101, 102...
> > 
> > * if it will be assumed that the issue / pr id will be followed by a slash or terminate immediately, it's also a wrong lead, as URLs ending with hashes are also submitted
> > ```
> > 
> > 
> > Working on a solution of the above problems.
> 
> What is the use case behind having urls with hashes? Can't we clean them in the add claim form?

We should surely add checks in the form where new claim is added, but the problem is, We already have entries like this.
image
image


So either we update the entries like these, or we accommodate things like this in the conflicts code. In any of these cases, we still need to add checks in the input form

**Comment by hereisnaman** [comment]
> > > A few thigns to mention:
> > > ```
> > > * if only the number is matched using a wildcard, then issue 10 will show conflicts with 100, 101, 102...
> > > 
> > > * if it will be assumed that the issue / pr id will be followed by a slash or terminate immediately, it's also a wrong lead, as URLs ending with hashes are also submitted
> > > ```
> > > 
> > > 
> > > Working on a solution of the above problems.
> > 
> > 
> > What is the use case behind having urls with hashes? Can't we clean them in the add claim form?
> 
> We should surely add checks in the form where new claim is added, but the problem is, We already have entries like this.
> image
> image
> 
> So either we update the entries like these, or we accommodate things like this in the conflicts code. In any of these cases, we still need to add checks in the input form

We can easily update those using a script. 

I'll increase the bounty if you include both the check in the form and add a script to do this.

**Comment by hereisnaman** [comment]
@YashKumarVerma can you first create a PR with the following:

- in add claim form, clean the urls, removing all the hashes, query params and add trailing slash.
- write a script to transform the current items for the same.

I'll give you separate points for that.

**Comment by YashKumarVerma** [comment]
> @YashKumarVerma can you first create a PR with the following:
> 
> * in add claim api, clean the urls, removing all the hashes, query params and add trailing slash.
> * write a script to transform the current items for the same.
> 
> I'll give you separate points for that.

Just one conformation, I've to do the first point on the server-side right ?

**Comment by YashKumarVerma** [comment]
With #390 sent, now we can create a more robust conflicts view.

**Comment by YashKumarVerma** [comment]
Carried forward in #392

**Pull request closed without merging**